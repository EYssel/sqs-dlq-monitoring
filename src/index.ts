import { Alarm, TreatMissingData } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import {
  EmailSubscription,
  LambdaSubscription,
} from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue, QueueProps } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';

export interface IMonitoredQueueProps {
  /** The properties of the SQS Queue Construct */
  readonly queueProps: QueueProps;
  /** The threshold for the amount of messages that are in the DLQ which trigger the alarm */
  readonly messageThreshold?: number;
  /** The threshold for the amount of messages that are in the DLQ which trigger the alarm */
  readonly evaluationThreshold?: number;
  /** The emails to which the messages should be sent */
  readonly emails?: string[];
  /** Slack bot token */
  readonly slackToken?: string;
  /** Slack channel to post messages to */
  readonly slackChannel?: string;
}

export class MonitoredQueue extends Construct {
  constructor(scope: Construct, id: string, props: IMonitoredQueueProps) {
    super(scope, id);

    const deadLetterQueue = props.queueProps.deadLetterQueue
      ? props.queueProps.deadLetterQueue
      : {
        queue: new Queue(this, 'DeadLetterQueue', {
          queueName: `${props.queueProps.queueName}-dlq`,
        }),
        maxReceiveCount: 3,
      };

    new Queue(this, 'Queue', {
      ...props.queueProps,
      deadLetterQueue,
    });

    const alarm = new Alarm(this, 'DLQ-Alarm', {
      metric: deadLetterQueue.queue.metricApproximateNumberOfMessagesVisible(),
      threshold: props.messageThreshold || 5,
      evaluationPeriods: props.evaluationThreshold || 1,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });

    const topic = new Topic(this, 'Topic', {
      topicName: `${deadLetterQueue.queue.queueName}-alarm-topic`,
    });

    const snsAction = new SnsAction(topic);

    alarm.addAlarmAction(snsAction);
    alarm.addOkAction(snsAction);

    props.emails
      ? this.addEmailNotificationDestination(topic, props.emails)
      : {};

    props.slackToken && props.slackChannel
      ? this.addSlackNotificationDestination(
        topic,
        props.slackToken,
        props.slackChannel,
      )
      : {};
  }

  addEmailNotificationDestination(topic: Topic, emails: string[]) {
    for (const email of emails) {
      topic.addSubscription(new EmailSubscription(email));
    }
  }

  addSlackNotificationDestination(
    topic: Topic,
    slackToken: string,
    slackChannel: string,
  ) {
    const slackListener = new NodejsFunction(this, 'SlackNotificationLambda', {
      entry: `${__dirname}/lambda/slackListener.ts`,
      environment: {
        SLACK_BOT_TOKEN: slackToken,
        SLACK_CHANNEL: slackChannel,
      },
    });

    topic.addSubscription(new LambdaSubscription(slackListener));
  }
}
