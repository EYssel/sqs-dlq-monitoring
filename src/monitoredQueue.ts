import { Alarm, TreatMissingData } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Architecture, Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';
import {
  EmailSubscription,
  LambdaSubscription,
} from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue, QueueProps } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';
import * as path from 'path';

export interface IMessagingProvider {
  deployProvider(scope: Construct, topic: Topic): void;
}

export class SlackProvider implements IMessagingProvider {
  /** Slack bot token for providing access to the Lambda function to write messages to Slack
   * @required
   */
  readonly slackToken: string;

  /** Slack channel to post messages to
   * @required
   */
  readonly slackChannel: string;

  /**
   * Unique name or identifier for the slackProvider
   */
  readonly name?: string;

  constructor(slackToken: string, slackChannel: string, name: string) {
    this.slackToken = slackToken;
    this.slackChannel = slackChannel;
    this.name = name;
  }

  deployProvider(scope: Construct, topic: Topic) {
    addSlackNotificationDestination(
      scope,
      topic,
      this.slackToken,
      this.slackChannel,
      this.name,
    );
  }
}

export class EmailProvider implements IMessagingProvider {
  /** The emails to which the messages should be sent */
  readonly emails: string[];

  constructor(emails: string[]) {
    this.emails = emails;
  }

  deployProvider(_scope: Construct, topic: Topic): void {
    addEmailNotificationDestination(topic, this.emails);
  }
}

export interface IMonitoredQueueProps {
  /** The standard properties of the SQS Queue Construct
   * @required
   */
  readonly queueProps: QueueProps;

  /** The number of times a message can be unsuccesfully dequeued before being moved to the dead-letter queue.
   * @default 3
   * @optional
   */
  readonly maxReceiveCount?: number;

  /** The threshold for the amount of messages that are in the DLQ which trigger the alarm
   * @default 5
   * @optional
   */
  readonly messageThreshold?: number;

  /** The number of periods over which data is compared to the specified threshold.
   * @default 1
   * @optional
   */
  readonly evaluationThreshold?: number;

  /**
   * A list of messaging providers that will be deployed and will listen for changes to the alarm.
   * @optional
   */
  readonly messagingProviders?: IMessagingProvider[];
}

export class MonitoredQueue extends Construct {
  constructor(scope: Construct, id: string, props: IMonitoredQueueProps) {
    super(scope, id);

    const deadLetterQueue = props.queueProps.deadLetterQueue
      ? props.queueProps.deadLetterQueue
      : {
        queue: new Queue(this, 'DeadLetterQueue', {
          queueName: `${props.queueProps.queueName}-dlq`,
        }),
        maxReceiveCount: props.maxReceiveCount || 3,
      };

    new Queue(this, 'Queue', {
      ...props.queueProps,
      deadLetterQueue,
    });

    const alarm = new Alarm(this, 'DLQ-Alarm', {
      alarmName: `${deadLetterQueue.queue.queueName}-alarm`,
      metric: deadLetterQueue.queue.metricApproximateNumberOfMessagesVisible(),
      threshold: props.messageThreshold || 5,
      evaluationPeriods: props.evaluationThreshold || 1,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });

    const topic = new Topic(this, 'Topic', {
      topicName: `${deadLetterQueue.queue.queueName}-alarm-topic`,
    });

    const snsAction = new SnsAction(topic);

    alarm.addAlarmAction(snsAction);
    alarm.addOkAction(snsAction);

    for (const messageProvider of props.messagingProviders
      ? props.messagingProviders
      : []) {
      messageProvider.deployProvider(this, topic);
    }
  }
}

function addEmailNotificationDestination(topic: Topic, emails: string[]) {
  for (const email of emails) {
    topic.addSubscription(new EmailSubscription(email));
  }
}

function addSlackNotificationDestination(
  scope: Construct,
  topic: Topic,
  slackToken: string,
  slackChannel: string,
  name?: string,
) {
  const slackListener = new Function(
    scope,
    'SlackListenerLambda' + (name ? `${name.toUpperCase()}` : randomIdentifier()),
    {
      runtime: Runtime.NODEJS_18_X,
      architecture: Architecture.ARM_64,
      code: Code.fromAsset(path.join(__dirname, '../lib/lambda/slackListener')),
      handler: 'index.handler',
      environment: {
        SLACK_BOT_TOKEN: slackToken,
        SLACK_CHANNEL: slackChannel,
      },
      logRetention: 7,
    },
  );

  topic.addSubscription(new LambdaSubscription(slackListener));
}

export function randomIdentifier() {
  return (Math.random() + 1).toString(36).substring(7).toUpperCase();
}
