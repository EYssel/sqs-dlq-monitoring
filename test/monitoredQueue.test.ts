import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { MonitoredQueue } from '../src/index';
import { EmailProvider, SlackProvider } from '../src/monitoredQueue';

describe('MonitoredQueue', () => {
  test('should create a monitored queue', () => {
    const stack = new Stack();
    new MonitoredQueue(stack, 'test', {
      queueProps: {
        queueName: 'test',
      },
    });

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('should create a monitored queue with a custom DLQ', () => {
    const stack = new Stack();
    new MonitoredQueue(stack, 'test', {
      queueProps: {
        queueName: 'test',
        deadLetterQueue: {
          queue: new Queue(stack, 'DLQ', {
            queueName: 'custom-dlq',
          }),
          maxReceiveCount: 3,
        },
      },
    });

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('should create a monitored queue with an email subscription', () => {
    const stack = new Stack();
    new MonitoredQueue(stack, 'test', {
      queueProps: {
        queueName: 'test',
      },
      messagingProviders: [
        new EmailProvider(['testemail@test.com']),
      ],
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('should create a monitored queue with a lambda SNS listener and Lambda Subscription', () => {
    const stack = new Stack();
    new MonitoredQueue(stack, 'test', {
      queueProps: {
        queueName: 'test',
      },
      messagingProviders: [
        new SlackProvider('test_token', 'test_channel'),
      ],
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
