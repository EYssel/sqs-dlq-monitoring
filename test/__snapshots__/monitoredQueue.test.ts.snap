// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitoredQueue should create a monitored queue 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testDLQAlarmFF94B263": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "testDeadLetterQueueA5F7D72D",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testDeadLetterQueueA5F7D72D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testQueueA1A031F3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "testDeadLetterQueueA5F7D72D",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testTopic20AA7407": {
      "Properties": {
        "TopicName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm-topic",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`MonitoredQueue should create a monitored queue with a custom DLQ 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DLQ581697C4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "custom-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testDLQAlarmFF94B263": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "DLQ581697C4",
                  "QueueName",
                ],
              },
              "-alarm",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "DLQ581697C4",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testQueueA1A031F3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "DLQ581697C4",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testTopic20AA7407": {
      "Properties": {
        "TopicName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "DLQ581697C4",
                  "QueueName",
                ],
              },
              "-alarm-topic",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`MonitoredQueue should create a monitored queue with a lambda SNS listener and Lambda Subscription 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testDLQAlarmFF94B263": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "testDeadLetterQueueA5F7D72D",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testDeadLetterQueueA5F7D72D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testQueueA1A031F3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "testDeadLetterQueueA5F7D72D",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testSlackNotificationLambdaAllowInvoketestTopicFC4454569C5983A0": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "testSlackNotificationLambdaF8C3E91E",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "testTopic20AA7407",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "testSlackNotificationLambdaF8C3E91E": {
      "DependsOn": [
        "testSlackNotificationLambdaServiceRole5E2AB417",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f28977b4f47f136211b3e3791f163fa3cef33786dc1f88f323bd6748fd9aa77a.zip",
        },
        "Environment": {
          "Variables": {
            "SLACK_BOT_TOKEN": "test",
            "SLACK_CHANNEL": "test",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "testSlackNotificationLambdaServiceRole5E2AB417",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testSlackNotificationLambdaLogRetention50573FD2": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "testSlackNotificationLambdaF8C3E91E",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "testSlackNotificationLambdaServiceRole5E2AB417": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testSlackNotificationLambdaTopicC5226703": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "testSlackNotificationLambdaF8C3E91E",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "testTopic20AA7407",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "testTopic20AA7407": {
      "Properties": {
        "TopicName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm-topic",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`MonitoredQueue should create a monitored queue with an email subscription 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testDLQAlarmFF94B263": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "testDeadLetterQueueA5F7D72D",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "OKActions": [
          {
            "Ref": "testTopic20AA7407",
          },
        ],
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testDeadLetterQueueA5F7D72D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test-dlq",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testQueueA1A031F3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "test",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "testDeadLetterQueueA5F7D72D",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testTopic20AA7407": {
      "Properties": {
        "TopicName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::GetAtt": [
                  "testDeadLetterQueueA5F7D72D",
                  "QueueName",
                ],
              },
              "-alarm-topic",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
    "testTopictestemailtestcomC7B96C79": {
      "Properties": {
        "Endpoint": "testemail@test.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "testTopic20AA7407",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
